### PR TITLE
'scope' can be applied to class member 'this'

### PR DESCRIPTION
**Pull request by WalterBright**
This supports the syntax for it.

The test file also apparently had a bunch of tabs that got converted to spaces by my git script.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
Looks good.

File: src/func.d [hunk 1]
```diff
@@ -733,8 +736,14 @@ extern (C++) class FuncDeclaration : Declaration
 
         if ((storage_class & STCauto) && !f.isref && !inferRetType)
             error("storage class 'auto' has no effect if return type is not inferred");
-        if (storage_class & STCscope)
+
+        /* Functions can only be 'scope' if they have a 'this' that is a pointer, not a ref
+         */
+        if (f.isscope && !isNested() &&
+            !(ad && ad.isClassDeclaration()))
```
Comment (andralex) [inline src/func.d]: at some point we should start taking advantage of no stinkin' trailing parens:

```d
if (f.isscope && !isNested &&  !(ad && ad.isClassDeclaration)
```

File: src/mtype.d [hunk 2]
```diff
@@ -5979,6 +5979,7 @@ extern (C++) final class TypeFunction : TypeNext
     bool isproperty;            // can be called without parentheses
     bool isref;                 // true: returns a reference
     bool isreturn;              // true: 'this' is returned by ref
+    bool isscope;               // true: 'this' is scope
```
Comment (andralex) [inline src/mtype.d]: these many bools start looking like a good candidate for `bitfields`...

**Comment by andralex** [comment]
Auto-merge toggled on

File: src/func.d (hunk 1 above)
Comment (WalterBright) [inline src/func.d]: I still like drawing a distinction between function calls and variable tests.

Comment (andralex) [inline src/func.d]: I know! It's like heroine!

File: src/mtype.d (hunk 2 above)
Comment (WalterBright) [inline src/mtype.d]: bitfields are in Phobos, which we aren't using for dmd at the moment.

**Pull request merged**